src/tangle/sources/inx.rs

```diff
@@ -1,2 +1,90 @@
+
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use futures::{stream::BoxStream, StreamExt, TryStreamExt};
+
+use super::{BlockData, InputSource, MilestoneData};
+use crate::{
+    inx::{Inx, InxError, MarkerMessage},
+    tangle::{ledger_updates::LedgerUpdateStore, milestone_range::MilestoneRange},
+    types::{
+        ledger::{LedgerSpent, MilestoneIndexTimestamp},
+        tangle::MilestoneIndex,
+    },
+};
+
+#[async_trait]
+impl InputSource for Inx {
+    type Error = InxError;
+
+    async fn milestone_stream(
+        &self,
+        range: MilestoneRange,
+    ) -> Result<BoxStream<Result<MilestoneData, Self::Error>>, Self::Error> {
+        let mut inx = self.clone();
+        Ok(Box::pin(
+            inx.listen_to_confirmed_milestones(range.into())
+                .await?
+                .and_then(|msg| async move {
+                    let payload = if let iota_types::block::payload::Payload::Milestone(payload) =
+                        msg.milestone.milestone.inner_unverified()?
+                    {
+                        payload.into()
+                    } else {
+                        // The raw data is guaranteed to contain a milestone payload.
+                        unreachable!();
+                    };
+                    Ok(MilestoneData {
+                        // TODO: What do we do here, enhance the error type?
+                        milestone_id: msg.milestone.milestone_info.milestone_id.unwrap(),
+                        at: MilestoneIndexTimestamp {
+                            milestone_index: msg.milestone.milestone_info.milestone_index,
+                            milestone_timestamp: msg.milestone.milestone_info.milestone_timestamp.into(),
+                        },
+                        payload,
+                        protocol_params: msg.current_protocol_parameters.params.inner_unverified()?.into(),
+                    })
+                }),
+        ))
+    }
+
+    async fn cone_stream(
+        &self,
+        index: MilestoneIndex,
+    ) -> Result<BoxStream<Result<BlockData, Self::Error>>, Self::Error> {
+        let mut inx = self.clone();
+        Ok(Box::pin(inx.read_milestone_cone(index.0.into()).await?.and_then(
+            |msg| async move {
+                Ok(BlockData {
+                    block_id: msg.metadata.block_id,
+                    block: msg.block.clone().inner_unverified()?.into(),
+                    raw: msg.block.data(),
+                    metadata: msg.metadata.into(),
+                })
+            },
+        )))
+    }
+
+    async fn ledger_updates(&self, index: MilestoneIndex) -> Result<LedgerUpdateStore, Self::Error> {
+        let mut inx = self.clone();
+        let mut stream = inx.listen_to_ledger_updates((index.0..=index.0).into()).await?;
+        let MarkerMessage {
+            consumed_count,
+            ..
+            // TODO: What do we do here?
+        } = stream.try_next().await?.unwrap().begin().unwrap();
```

Comment:
I think we should rip everything out of the `inx` module and put it into this module / impl here. We should then extend the `InxError` type with the variants that are required by our new implementation.

Reply:
Alright, maybe we should create an issue then to track that progress.

Reply:
Sounds good, can you open the issue please?